mlserver/batching/requests.py

```diff
@@ -0,0 +1,165 @@
+
+from ..types import (
+    InferenceRequest,
+    InferenceResponse,
+    Parameters,
+    RequestInput,
+    ResponseOutput,
+)
+
+
+def _get_data(payload: Union[RequestInput, ResponseOutput]):
+    return getattr(payload.data, "__root__", payload.data)
+
+
+def _infer_elem_size(response_output: ResponseOutput) -> int:
+    batch_axis = _get_batch_axis(response_output)
+    shape_without_batch = response_output.shape.copy()
+    shape_without_batch.pop(batch_axis)
+    return reduce(mul, shape_without_batch, 1)
+
+
+def _get_batch_size(request_input: RequestInput) -> int:
+    batch_dim = _get_batch_axis(request_input)
+    return request_input.shape[batch_dim]
+
+
+def _get_batch_axis(
+    request_input: Union[RequestInput, ResponseOutput]  # type: ignore
+) -> int:
+    # TODO: Support other batch dimensions
+    return 0
+
+
+class BatchedRequests:
+    def __init__(self, inference_requests: List[InferenceRequest] = []):
+        self._inference_requests = inference_requests
+        self._prediction_ids: List[str] = [  # type: ignore
+            req.id for req in self._inference_requests  # type: ignore
+        ]
+        self._minibatch_sizes: List[int] = []
+        self.merged_request = self._merge_requests()
+
+    def _merge_requests(self) -> InferenceRequest:
+        inputs_index: Dict[str, List[RequestInput]] = defaultdict(list)
+        all_params = {}
+
+        for inference_request in self._inference_requests:
+            if inference_request.parameters:
+                request_params = inference_request.parameters.dict(exclude_unset=True)
+                all_params.update(request_params)
+            self._minibatch_sizes = []
+            for request_input in inference_request.inputs:
+                inputs_index[request_input.name].append(request_input)
+
+        inputs = [
+            self._merge_request_inputs(request_inputs)
+            for request_inputs in inputs_index.values()
+        ]
+
+        # TODO: Add outputs
+        # TODO: Should we add a 'fake' request ID?
+        params = Parameters(**all_params) if all_params else None
+        return InferenceRequest(inputs=inputs, parameters=params)
+
+    def _merge_request_inputs(self, request_inputs: List[RequestInput]) -> RequestInput:
+        batch_size, data, parameters = self._merge_data(request_inputs)
+
+        # TODO: What should we do if list is empty?
+        sampled = request_inputs[0]
+
+        batch_axis = _get_batch_axis(sampled)
+        shape = sampled.shape.copy()
+        shape.pop(batch_axis)
+        shape.insert(batch_axis, batch_size)
+
+        return RequestInput(
+            name=sampled.name,
+            datatype=sampled.datatype,
+            shape=shape,
+            data=data,
+            parameters=parameters,
+        )
+
+    def _merge_data(
+        self, request_inputs: List[RequestInput]
+    ) -> Tuple[int, Any, Optional[Parameters]]:
+        # We assume all inputs will have the same minibatch size
```

Comment:
is this comment still true?

Reply:
It is, across different input heads. That would a 2nd-level type of variability, e.g. we've got 2 input heads (e.g. `words` and `images`), but we get 1 image for each 3 words. So we'd have to workout how to map each cardinality (e.g. 3 and 1) with the number of output tensors sent back. 

To simplify things, for now we assume that cardinality across inputs will always be the same. That is, there will be the same number of elements on each input head.

Reply:
should there be an assertion / exception to be the case, if not already done?